routers/web/repo/compare.go

```diff
@@ -395,15 +398,26 @@ func ParseCompareInfo(ctx *context.Context) (*models.User, *models.Repository, *
 	if rootRepo != nil &&
 		rootRepo.ID != headRepo.ID &&
 		rootRepo.ID != baseRepo.ID {
-		perm, branches, tags, err := getBranchesAndTagsForRepo(ctx.User, rootRepo)
-		if err != nil {
-			ctx.ServerError("GetBranchesForRepo", err)
-			return nil, nil, nil, nil, "", ""
-		}
-		if perm {
-			ctx.Data["RootRepo"] = rootRepo
-			ctx.Data["RootRepoBranches"] = branches
-			ctx.Data["RootRepoTags"] = tags
+		if !fileOnly {
+			perm, branches, tags, err := getBranchesAndTagsForRepo(ctx.User, rootRepo)
+			if err != nil {
+				ctx.ServerError("GetBranchesForRepo", err)
+				return nil, nil, nil, nil, "", ""
+			}
+			if perm {
+				ctx.Data["RootRepo"] = rootRepo
+				ctx.Data["RootRepoBranches"] = branches
+				ctx.Data["RootRepoTags"] = tags
+			}
+		} else {
+			perm, err := models.GetUserRepoPermission(rootRepo, ctx.User)
```

Comment:
I see the permission check appears in many places. Maybe we can introduce a common function. And move the user permission check away from `getBranchesAndTagsForRepo` also.

Reply:
it's a different repo each time. (or at least it should be.)